### ISSUE TITLE
Road to the 1.0.0

### ISSUE THREAD

**Issue by Swiip** [Swiip/generator-gulp-angular#259]
Hi everyone,

We talk a lot with @zckrs about the design of the 1.0.0 but we have to communicate with everyone what we think.

We now have all the main features in the generator. I think that no more major feature will be added before the release. There is some exception of feature we still want to add, but it should be limited. All features of the generator add complexity and a charge to maintain it so at this point we have to stop add things.

In the master branch, there is already a big refactoring on the gulp files. We planned also a refactoring of the source files and the test files in order to have a source base more readable and maintainable.

First of all, we have to limit big feature PRs in order to be able to do the refactoring. After, we'll have to focus on bugs. I think we'll use the 0.9 version soon as a pre 1.0.

We'll have also to upgrade a little the project presentation in the readme and the Wiki.

After the 1.0.0 release, it's possible we keep the feature lock. Indeed, we can't continue to add complexity without a limit. The generator would be too generic and not useful anymore. Perhaps we'll start to work on generator composition...

We're still open to discussion about this choices.

## Remaining for the 1.0.0

- [x]  Organize / split `app/src/format.js`
- [x] Organize / split / complete `test/test-files-generate.mocha.js`
- [x] Update es6 / browserify processing

- [x] Fix misc bug
- [x] Update `README.md`
- [x] Illustrate a directive in sample app https://github.com/Swiip/generator-gulp-angular/issues/261#issuecomment-81790092
- [x] Improve ES6 template
- [x] Wiki page "How it works"
- [x] New design for the generated website

## Tasks postponed after the 1.0.0

- Component generator
- Script loader feature
- Test framework choice
- Updating to Gulp 4

**Comment by zckrs** [comment]
:+1:

**Comment by Toilal** [comment]
Please merge my PRs before starting the refactoring, those are quite bigs :)

After those merge, i don't see any features missing, it's complete and already a great generator !

Something that should be done during the refactoring is having an easier way to manage global dependencies used by the generator. Maintaining those dependencies is quite complex for now, so here's my idea :

- Declare all possible dependencies for all features in 2 new files : `package.json` and `bower.json`.
- Create a new `dependencies.js` file that will require those central `package.json` and `bower.json` to build a list of dependencies that are required, based on selected options.
- Use those lists to generate `package.json` and `bower.json`.

With this workflow, it will be easy to update or add a dependency to the generator, because dependencies declared in central files `package.json` and `bower.json` can be updated with tools like [npm-check-updates](https://www.npmjs.com/package/npm-check-updates) and [bower-update](https://www.npmjs.com/package/bower-update).

It will also avoid to manually maintain the `/test/deps/bower.json` and '/test/deps/package.json', as central files could be used directly for this purpose.

**Comment by Swiip** [comment]
We have to clean up the current PRs before further refactoring yes.

A better solution for handling deps seams necessary. I've my own idea on the matter, I'll describe it soon.

**Comment by marani** [comment]
- component generator, perhaps this could be solved with composability - using `generator-angular`
- Lazy loading, or entry-point based module bundling (integrating gulp-webpack/requirejs/angularAMD).

What do you think about these features after 1.0 ?

**Comment by zckrs** [comment]
* rewrite README
* rewrite test
* a new design to celebrate v1.0.0

**Comment by Swiip** [comment]
I'll edit the issue with the remaining of the 1.0 and the tasks identified for after the release. It's not final, if we find other tasks to do before the release, I'll update the list.

**Comment by Swiip** [comment]
Remains the question of Gulp 4. Will it ready after or before us? :trollface:

**Comment by zckrs** [comment]
Gulp4 is not yet official.
I start to rewrite our gulpfile but this is HUGE. 

I think we should do that after v1.0.0 and maybe Gulp4 should be released

**Comment by Swiip** [comment]
Yeah, it's not released yet, and I agree it seams difficult for the 1.0.

It's just that it could end up weird if we release a 1.0 on Gulp 3 just after the release of the 4.

**Comment by zckrs** [comment]
Anyway released before or after our v1.0.0 we will create a branch `gulp4`

**Comment by Swiip** [comment]
I think that a new wiki page could be useful, something like "how it works".

It would describe the main processes that the gulp build use. I think it could help resolving a log of issues.

**Comment by zckrs** [comment]
:+1:

**Comment by jamm** [comment]
I thankful for ES6 support but I really dislike I need to write 4 times declaration of controller and 2 times list of dependencies. Any code duplication is evil.
Example:
1. `class ActivityCtrl `
2. `export default ActivityCtrl;`
3. `import ActivityCtrl from './activity/ActivityCtrl.js';`
4. `.controller('ActivityCtrl', ActivityCtrl)`
dependencies:

p1.

```js
class ActivityCtrl {
  constructor(UserEventsLog) {
```

p2. `ActivityCtrl.$inject = ['UserEventsLog'];`

**Comment by Toilal** [comment]
@jamm maybe you should open another issue, but it's an interesting point.

**Comment by Swiip** [comment]
@jamm I totally agree with you. I looked through some example implementations and didn't find nothing really better.

The Angular 1 module system and the ES6 module system are definitely in conflict... Any better solution is welcomed.

**Comment by jamm** [comment]
@Swiip we still can declare them old-fashioned way:
angular.module('appName').service[controller]('serviceName', function dependenciesHere($http){
  class es6ModernClass{
  }
 return new es6ModernClass();
});

It works. Example from my experiments: https://github.com/jamm/AquilamusTest/blob/master/app/es6/services/materialsservice.js

**Comment by Swiip** [comment]
Yeah, I thought about that. But does es6 without modules still es6?

**Comment by jamm** [comment]
@Swiip it's not "true" modules anyway, they all are concatenated in one file :) I think in Angular 2 we will see some true modules when we will load different modules for different parts of app.
I like that we don't need to write name of app in existing variant, but duplicated list of dependencies is erroneous thing, so I (personally) ready to pay this price for less bugs in code. Not sure if other users will support me, but hope so :)

**Comment by Toilal** [comment]
I think it's better to not use ES6 modules as AngularJS already has modules. There's plenty of things in ES6 to have fun without having to force the usage of module, those doesn't provide anything more using angularJS 1.x.

That beeing said, the generator only gives a template, and user can already write the @jamm way at the moment.

**Comment by Swiip** [comment]
It's complicated, there is no perfect code. We can't deliver a bootstrap application which match everyone code styles.

If your arguments match with me, I'm pretty sure that if I remove modules support from the generator it will be asked really soon... I don't want to end up maintaining multiples default apps.

**Comment by jamm** [comment]
@Toilal currently in index.html I don't see any includes of apps files, so I'm not sure I can use it without imports. But will try. Too many things for experiments :)

**Comment by Toilal** [comment]
About refactoring, there's many `'none'` string values and related checking in the project. Why not just using the `undefined` value of JavaScript ?

**Comment by Toilal** [comment]
@jamm `src/index.html` won't get modified by injection, you have to read the `.tmp/server/index.html` file after running the inject gulp task (or `gulp serve`).

**Comment by jamm** [comment]
@Toilal after experiments I found that if I add  `<script src="pathToJsFile"></script>` into `<!-- build:js({.tmp,src}) scripts/app.js -->` but not inside `<!-- inject:js -->`, then everything works fine.

**Comment by Toilal** [comment]
It's normal, content between  tag is replaced by gulp with dependencies.

But you should not add <script src="..."> tags, this should be handled automatically by gulp, unless your script is not coming from bower.

**Comment by jamm** [comment]
In [comment](https://plus.google.com/u/0/+PascalPrecht/posts/Jre92W8GnJQ) from Christoph Burgdorf I found interesting idea that constructor() being parsed as $inject so we don't need to call .$inject. I tested it and it works! 
@Swiip , @zckrs , we really don't need `className.$inject = ['services']`, constructor is enough :)

**Comment by jamm** [comment]
my previous comment was wrong... it works only before minification...

**Comment by jamm** [comment]
and another one my assumption were wrong. We can't use index.html to include custom js files of our app to avoid export/import/declaration boilerplate code, if we use es6 features in these files. Because they will not be converted to es5.
So current configuration of ES6 is only possible. Sorry all for noise.

**Comment by jamm** [comment]
correction: only possible with current gulp config. If you could make gulp convert included js files too - it would be cool. But I don't know if it's possible at all.

**Comment by Swiip** [comment]
Ok, I see the problem, I had to work on the question to find the best way to handle that.

**Comment by jamm** [comment]
See what they are doing, switching from bower to npm: https://github.com/tastejs/todomvc/issues/1110

**Comment by Swiip** [comment]
I'm considering using NPM. I think I first will remove the Bower logo from the Readme. It's the current choice but it's not tied with the generator.

At first, I was thinking that NPM will be a need when and only when I'll address the script loading part and Browserify. Now, I discover than Bower/NPM could be an independent option.

But I think it's a too big change for the 1.0 but perhaps just after.

I'm just a little afraid because Bower libs are just starting to be precise with targets: js file(s) but also css, sass, or fonts. It just starting to works well with wiredep. But unless I'm mistaking, this is not possible with NPM.

**Comment by jamm** [comment]
I know I sound boring and annoying, sorry, but @stryju has posted solution of ugly .$inject:
https://gist.github.com/stryju/77109a3e1af24304ad36#file-foo-controller-es6-L2-L5
I've tested it and it works :)

**Comment by Swiip** [comment]
It's not boring, it's actually a subject I'm interesting at: how to makes Angular 1 and ES6 mix well. I'm fully open to a better implementation of the ES6 code ;)

**Comment by stryju** [comment]
i've been mixing the two with fair share of success (production-grade, but can't reveal it yet :wink:) - if you need some help let me know

**Comment by williamboman** [comment]
How comprehensive would it be to implement a component generator?

**Comment by intellix** [comment]
Personally, I've never seen the point in any of the component generators.

How is it any less effort than using snippets like: https://github.com/alsfurlan/ng-snippets, typing `ngcontroller` and press TAB? The files are automatically injected into your index.html anyway

**Comment by jamm** [comment]
@intellix generator of route can generate 4 files (controller, template, e2e test, unit-test), add route and imports into index.js (app.js), it's ~15 minutes of saved time. Not a big deal, but very handy. 
But everything has a price, and I think generator for so many supported JS dialects can be tricky to implement. Although, with some abstractions layers, it's possible, in long term.

**Comment by jamm** [comment]
About `// @ngInject` instead of `.$inject = []`:
babel doesn't work with code, generated by ng-annotate, but it can be resolved by adding 2 linea into build.js:

1) in the beginning:
`var replace = require('gulp-replace');`
2) into `html` task pipeline, after ngAnnotate:
```
      .pipe(replace('var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };', 'var _classCallCheck = function (){};'))
```
it just removes unnecessary check. Kind of hack, but works fine.

**Comment by zckrs** [comment]
v1.0.0 incoming

**Comment by Swiip** [comment]
1.0.0-rc1 is out, it's time to check if everything's ok.

We have a lot of suggestions in the issues. We loved to please everything but we simply can't do everything.

To pass to the 1.0 tag, I'll will close some of them using our label `good-but-no-time` which means that it seams great but it's not planned for a close future and we need to get our issues list clear. Any PR on these feature are welcome.

There will be a specific category of them which will be `gulp 4` which are specific questions that we choose to not answer with the 1.0 and gulp 3. I don't to have ugly fix just before the 1.0 when it will be clean and shiny with gulp 4.

I don't have a planning but it should be a gulp 4 branch somewhere after the 1.0 final.

**Comment by Swiip** [comment]
And.... Here we go ! 

https://github.com/Swiip/generator-gulp-angular/releases/tag/v1.0.0
